### PR TITLE
UCP/PROTO: Make perf nodes names more informative

### PR DESCRIPTION
**Pull request by ivankochin**
## What
Place full protocol chain to the perf nodes' names.

## Why ?
Makes protocol graph debugging more convenient.

| Before| After|
|--------|--------|
| image| image|

### CODE REVIEW

**Comment by yosefe** [comment]
do we really need it? if we look at the child (or descendant) nodes of a protocol node, we could see the protocol chain

**Comment by ivankochin** [comment]
> do we really need it? if we look at the child (or descendant) nodes of a protocol node, we could see the protocol chain

There is no strong need for this improvement, but FMPOV the line of "RTS" labels at the top levels looks ugly and uninformative. One who wants just to understand what particular proto was chosen in some particular case needs to dig deeper down to the graph to find the node with the protocol name (which even doesn't highlighted).

And with that patch top level contains the full chain of protocols which help to understand the whole picture.

**Comment by brminich** [comment]
> > do we really need it? if we look at the child (or descendant) nodes of a protocol node, we could see the protocol chain
> 
> There is no strong need for this improvement, but FMPOV the line of "RTS" labels at the top levels looks ugly and uninformative. One who wants just to understand what particular proto was chosen in some particular case needs to dig deeper down to the graph to find the node with the protocol name (which even doesn't highlighted).
> 
> And with that patch top level contains the full chain of protocols which help to understand the whole picture.

i tend to agree that it looks better than existing naming

**Comment by yosefe** [comment]
Maybe there is a way to make the description more compact, so the size of the tree node will not increase?

**Review by yosefe — COMMENTED** [review]
in general LGTM

File: src/ucp/proto/proto_debug.c [hunk 1]
```diff
@@ -512,9 +514,13 @@ ucp_proto_perf_node_new(ucp_proto_perf_node_type_t type, const char *name,
     }
 
     perf_node->type     = type;
-    perf_node->name     = name;
     perf_node->refcount = 1;
     ucs_array_init_dynamic(&perf_node->children);
+
+    if (name == NULL) {
+        name = null_string;
```
[the remark this answers is not in the document]
Comment (ivankochin) [reply src/ucp/proto/proto_debug.c:521]: Do you mean the following change?
```suggestion
        name = "(null)";
```

File: src/ucp/proto/proto_debug.h [hunk 2]
```diff
@@ -10,6 +10,9 @@
 #include "proto_select.h"
 
 
+/* String to place new line inside performance graph */
+#define UCP_PROTO_PERF_NODE_NEW_LINE "<BR/>"
```
[the remark this answers is not in the document]
Comment (ivankochin) [reply src/ucp/proto/proto_debug.h:14]: Do you mean the following change?
```suggestion
#define ucp_proto_perf_node_new_line "<BR/>"
```

You are right that is used only once, but I think it worth to create macro for that for code self-documentation purpose

File: src/ucp/proto/proto_debug.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (yosefe) [reply src/ucp/proto/proto_debug.c:521]: yes

File: src/ucp/proto/proto.h [hunk 3]
```diff
@@ -33,7 +33,6 @@
 /* Threshold for considering two performance values as equal */
 #define UCP_PROTO_PERF_EPSILON     1e-15
```
Comment (ivankochin) [inline src/ucp/proto/proto.h:35]:
```suggestion


```

**Review by brminich — APPROVED** [review]

File: src/ucp/proto/proto_debug.c [hunk 4]
```diff
@@ -512,9 +512,13 @@ ucp_proto_perf_node_new(ucp_proto_perf_node_type_t type, const char *name,
     }
 
     perf_node->type     = type;
-    perf_node->name     = name;
     perf_node->refcount = 1;
     ucs_array_init_dynamic(&perf_node->children);
+
+    if (name == NULL) {
```
Comment (yosefe) [inline src/ucp/proto/proto_debug.c:518]: in which flow the name can be NULL?

File: src/ucp/proto/proto_debug.c [hunk 5]
```diff
@@ -512,9 +512,13 @@ ucp_proto_perf_node_new(ucp_proto_perf_node_type_t type, const char *name,
...
+        name = "(null)";
+    }
```
Comment (ivankochin) [inline src/ucp/proto/proto_debug.c:520]:
```suggestion
    ucs_assert(name != NULL);
```

File: src/ucp/proto/proto_debug.c (hunk 4 above)
Comment (ivankochin) [reply src/ucp/proto/proto_debug.c]: Changed to assert

Comment (ivankochin) [reply src/ucp/proto/proto_debug.c]: Answering you question: `rndv/rkey_ptr` proto set `params.send_op` to `UCT_EP_OP_LAST` which causes setting `uct_ep_operation_names[UCT_EP_OP_LAST]` which is NULL to name of the `ucp_proto_common_lane_perf_node` result node.

**Comment by ivankochin** [comment]
@yosefe pls review

File: src/uct/base/uct_iface.c [hunk 6]
```diff
@@ -39,7 +39,7 @@ const char *uct_ep_operation_names[] = {
     [UCT_EP_OP_RNDV_ZCOPY]   = "rndv_zcopy",
     [UCT_EP_OP_ATOMIC_POST]  = "atomic_post",
     [UCT_EP_OP_ATOMIC_FETCH] = "atomic_fetch",
-    [UCT_EP_OP_LAST]         = NULL
+    [UCT_EP_OP_LAST]         = "unspecified"
```
Comment (yosefe) [inline src/uct/base/uct_iface.c:42]: maybe we should not create a perf_node when op==UCT_EP_OP_LAST? i.e skip calling ucp_proto_perf_node_new in this case, when name is supposedly NULL

Comment (ivankochin) [reply src/uct/base/uct_iface.c]: Done.

**Review by yosefe — APPROVED** [review]

**Pull request merged**